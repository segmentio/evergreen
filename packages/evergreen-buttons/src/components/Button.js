import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { Text } from 'evergreen-typography'
import { IconMap, IconAim } from 'evergreen-icons'
import {
  getBorderRadiusForControlHeight,
  getTextStyleForControlHeight,
  getIconSizeForControlHeight
} from 'evergreen-shared-styles'
import ButtonAppearances from '../styles/ButtonAppearances'

export default class Button extends PureComponent {
  static propTypes = {
    ...Text.propTypes,
    appearance: PropTypes.oneOf(Object.keys(ButtonAppearances)).isRequired,
    isActive: PropTypes.bool,

    // Icon support
    iconBefore: PropTypes.oneOf(Object.keys(IconMap)),
    iconBeforeAim: PropTypes.oneOf(Object.keys(IconAim)),
    iconAfter: PropTypes.oneOf(Object.keys(IconMap)),
    iconAfterAim: PropTypes.oneOf(Object.keys(IconAim))
  }

  static defaultProps = {
    is: 'button',
    position: 'relative',
    appearance: 'default',
    paddingTop: 0,
    paddingBottom: 0,
    height: 32,
    fontFamily: 'ui',
    fontWeight: 500,
    display: 'inline-flex',
    alignItems: 'center',
    flexWrap: 'nowrap',

    isActive: false,
    iconBeforeAim: 'none',
    iconAfterAim: 'none'
  }

  render() {
    const {
      css,
      height,
      isActive,
      children,
      appearance,

      // Paddings
      paddingRight,
      paddingLeft,
      paddingTop,
      paddingBottom,

      // Icons
      iconBefore: iconBeforeKey,
      iconBeforeAim,
      iconAfter: iconAfterKey,
      iconAfterAim,

      ...props
    } = this.props
    const appearanceStyle = ButtonAppearances[appearance]
    const textStyle = getTextStyleForControlHeight({ height })
    const borderRadius = getBorderRadiusForControlHeight({ height })
    const iconHeight = height - 4
    const iconSize = getIconSizeForControlHeight({ height: iconHeight })

    const pr = paddingRight ? paddingRight : Math.round(height / 2)
    const pl = paddingLeft ? paddingLeft : Math.round(height / 2)

    let iconBefore
    if (iconBeforeKey) {
      iconBefore = React.createElement(IconMap[iconBeforeKey], {
        aim: iconBeforeAim,
        iconSize,
        color: 'inherit',
        size: iconHeight,
        marginLeft: -Math.round(pl * 0.6)
      })
    }

    let iconAfter
    if (iconAfterKey) {
      iconAfter = React.createElement(IconMap[iconAfterKey], {
        aim: iconAfterAim,
        iconSize,
        color: 'inherit',
        size: iconHeight,
        marginRight: -Math.round(pl * 0.6)
      })
    }

    return (
      <Text
        borderTopRightRadius={borderRadius}
        borderBottomRightRadius={borderRadius}
        borderTopLeftRadius={borderRadius}
        borderBottomLeftRadius={borderRadius}
        paddingTop={paddingTop}
        paddingBottom={paddingBottom}
        paddingRight={pr}
        paddingLeft={pl}
        margin={0} // Removes weird margins in Safari
        {...textStyle}
        css={{
          ...css,
          ...appearanceStyle
        }}
        height={height}
        lineHeight={`${height}px`}
        {...(isActive ? { 'data-active': true } : {})}
        {...props}
      >
        {iconBefore || null}
        <span>{children}</span>
        {iconAfter || null}
      </Text>
    )
  }
}
