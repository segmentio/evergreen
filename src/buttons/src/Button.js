import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { Text } from '../../typography'
import { IconMap, IconAim } from '../../icons'
import {
  getBorderRadiusForControlHeight,
  getTextStyleForControlHeight,
  getIconSizeForControlHeight
} from '../../shared-styles'
import { Spinner } from '../../spinner'
import ButtonAppearances from './styles/ButtonAppearances'

export default class Button extends PureComponent {
  static propTypes = {
    /**
     * Composes the Text component as the base.
     */
    ...Text.propTypes,

    /**
     * The appearance of the button.
     */
    appearance: PropTypes.oneOf(Object.keys(ButtonAppearances)).isRequired,

    /**
     * When true, show a loading spinner before the children.
     * This also disables the button.
     */
    isLoading: PropTypes.bool,

    /**
     * Forcefully set the active state of a button.
     * Useful in conjuction with a Popover.
     */
    isActive: PropTypes.bool,

    /**
     * Sets an icon before the text. Can be any icon from Evergreen.
     */
    iconBefore: PropTypes.oneOf(Object.keys(IconMap)),

    /**
     * The aim of the left icon. Not a big use case for this.
     */
    iconBeforeAim: PropTypes.oneOf(Object.keys(IconAim)),

    /**
     * Sets an icon after the text. Can be any icon from Evergreen.
     */
    iconAfter: PropTypes.oneOf(Object.keys(IconMap)),

    /**
     * The aim of the right icon. Useful to aim a triangle down.
     */
    iconAfterAim: PropTypes.oneOf(Object.keys(IconAim)),

    /**
     * When true, the button is disabled.
     * isLoading also sets the button to disabled.
     */
    disabled: PropTypes.bool
  }

  static defaultProps = {
    appearance: 'default',
    isActive: false,
    iconBeforeAim: 'none',
    iconAfterAim: 'none',
    height: 32,
    paddingTop: 0,
    paddingBottom: 0
  }

  static styles = {
    position: 'relative',
    fontFamily: 'ui',
    fontWeight: 500,
    display: 'inline-flex',
    alignItems: 'center',
    flexWrap: 'nowrap'
  }

  render() {
    const {
      css,
      height,
      isActive,
      children,
      disabled,
      appearance,
      isLoading,

      // Paddings
      paddingRight,
      paddingLeft,
      paddingTop,
      paddingBottom,

      // Icons
      iconBefore: iconBeforeKey,
      iconBeforeAim,
      iconAfter: iconAfterKey,
      iconAfterAim,

      ...props
    } = this.props
    const appearanceStyle = ButtonAppearances[appearance]
    const textStyle = getTextStyleForControlHeight({ height })
    const borderRadius = getBorderRadiusForControlHeight({ height })
    const iconHeight = height - 4
    const iconSize = getIconSizeForControlHeight({ height: iconHeight })

    const pr =
      paddingRight !== undefined ? paddingRight : Math.round(height / 2) // eslint-disable-line no-negated-condition
    const pl = paddingLeft !== undefined ? paddingLeft : Math.round(height / 2) // eslint-disable-line no-negated-condition

    let iconBefore
    if (iconBeforeKey) {
      iconBefore = React.createElement(IconMap[iconBeforeKey], {
        aim: iconBeforeAim,
        iconSize,
        color: 'inherit',
        size: iconHeight,
        marginLeft: -Math.round(pl * 0.6)
      })
    }

    let iconAfter
    if (iconAfterKey) {
      iconAfter = React.createElement(IconMap[iconAfterKey], {
        aim: iconAfterAim,
        iconSize,
        color: 'inherit',
        size: iconHeight,
        marginRight: -Math.round(pl * 0.6)
      })
    }

    return (
      <Text
        is="button"
        borderTopRightRadius={borderRadius}
        borderBottomRightRadius={borderRadius}
        borderTopLeftRadius={borderRadius}
        borderBottomLeftRadius={borderRadius}
        paddingTop={paddingTop}
        paddingBottom={paddingBottom}
        paddingRight={pr}
        paddingLeft={pl}
        margin={0} // Removes weird margins in Safari
        {...textStyle}
        color={null} // Prevent the Text color overriding the glamor appearanceStyle color
        css={{
          ...appearanceStyle,
          ...css
        }}
        height={height}
        lineHeight={`${height}px`}
        {...(isActive ? { 'data-active': true } : {})}
        {...Button.styles}
        {...props}
        disabled={disabled || isLoading}
      >
        {isLoading && (
          <Spinner
            marginLeft={-Math.round(height / 8)}
            marginRight={Math.round(height / 4)}
            size={Math.round(height / 2)}
          />
        )}
        {iconBefore || null}
        <span>{children}</span>
        {iconAfter || null}
      </Text>
    )
  }
}
