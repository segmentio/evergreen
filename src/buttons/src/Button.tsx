import React, { PureComponent, ReactNode } from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import { Text } from '../../typography'
import { Icon, IconName } from '../../icon'
import { Spinner } from '../../spinner'
import { withTheme, Theme } from '../../theme'

type Intent = 'none' | 'success' | 'warning' | 'danger'
type Appearance = 'default' | 'minimal' | 'primary'

interface ButtonProps extends React.ComponentProps<typeof Text> {
  /**
   * The intent of the button.
   */
  intent: Intent

  /**
   * The appearance of the button.
   */
  appearance: Appearance

  /**
   * Sets the height and font-size of the button.
   */
  height: number

  /**
   * When true, show a loading spinner before the children.
   * This also disables the button.
   */
  isLoading?: boolean

  /**
   * Forcefully set the active state of a button.
   * Useful in conjuction with a Popover.
   */
  isActive?: boolean

  /**
   * Sets an icon before the text. Can be any icon from Evergreen.
   */
  iconBefore?: IconName

  /**
   * Sets an icon after the text. Can be any icon from Evergreen.
   */
  iconAfter?: IconName

  /**
   * When true, the button is disabled.
   * isLoading also sets the button to disabled.
   */
  disabled?: boolean

  /**
   * Theme provided by ThemeProvider.
   */
  theme: Theme
}

class Button extends PureComponent<ButtonProps> {
  static propTypes = {
    intent: PropTypes.oneOf(['none', 'success', 'warning', 'danger'])
      .isRequired as PropTypes.Validator<Intent>,
    appearance: PropTypes.oneOf(['default', 'minimal', 'primary'])
      .isRequired as PropTypes.Validator<Appearance>,
    height: PropTypes.number.isRequired,
    isLoading: PropTypes.bool,
    isActive: PropTypes.bool,
    iconBefore: PropTypes.string as PropTypes.Requireable<IconName>,
    iconAfter: PropTypes.string as PropTypes.Requireable<IconName>,
    disabled: PropTypes.bool,
    theme: PropTypes.object.isRequired as PropTypes.Validator<Theme>
  }

  static defaultProps = {
    appearance: 'default' as const,
    height: 32,
    intent: 'none' as const,
    isActive: false,
    paddingBottom: 0,
    paddingTop: 0
  }

  static styles = {
    position: 'relative',
    fontFamily: 'ui',
    fontWeight: 500,
    display: 'inline-flex',
    alignItems: 'center',
    flexWrap: 'nowrap'
  }

  render() {
    const {
      theme,
      className,

      intent,
      height,
      isActive,
      children,
      disabled,
      appearance,
      isLoading,

      // Paddings
      paddingRight,
      paddingLeft,
      paddingTop,
      paddingBottom,

      // Icons
      iconBefore: iconBeforeKey,
      iconAfter: iconAfterKey,

      ...props
    } = this.props

    const themedClassName = theme.getButtonClassName(appearance, intent)
    const textSize = theme.getTextSizeForControlHeight(height)

    const borderRadius = theme.getBorderRadiusForControlHeight(height)
    const iconSize = theme.getIconSizeForButton(height)

    const pr =
      typeof paddingRight === 'number' ? paddingRight : Math.round(height / 2)
    const pl =
      typeof paddingLeft === 'number' ? paddingLeft : Math.round(height / 2)

    let iconBefore: ReactNode | undefined
    if (iconBeforeKey) {
      iconBefore = (
        <Icon
          icon={iconBeforeKey}
          size={iconSize}
          marginLeft={-Math.round(pl * 0.2)}
          marginRight={Math.round(iconSize * 0.7)}
        />
      )
    }

    let iconAfter: ReactNode | undefined
    if (iconAfterKey) {
      iconAfter = (
        <Icon
          icon={iconAfterKey}
          size={iconSize}
          marginRight={-Math.round(pl * 0.2)}
          marginLeft={Math.round(iconSize * 0.7)}
        />
      )
    }

    return (
      <Text
        is="button"
        className={cx(themedClassName, className)}
        borderTopRightRadius={borderRadius}
        borderBottomRightRadius={borderRadius}
        borderTopLeftRadius={borderRadius}
        borderBottomLeftRadius={borderRadius}
        paddingTop={paddingTop}
        paddingBottom={paddingBottom}
        paddingRight={pr}
        paddingLeft={pl}
        marginLeft={0} // Removes weird margins in Safari
        marginRight={0} // Removes weird margins in Safari
        marginTop={0} // Removes weird margins in Safari
        marginBottom={0} // Removes weird margins in Safari
        size={textSize}
        color={null} // Prevent the Text color overriding the glamor appearanceStyle color
        height={height}
        lineHeight={`${height}px`}
        {...(isActive ? { 'data-active': true } : {})}
        {...Button.styles}
        {...props}
        disabled={disabled || isLoading}
      >
        {isLoading && (
          <Spinner
            marginLeft={-Math.round(height / 8)}
            marginRight={Math.round(height / 4)}
            size={Math.round(height / 2)}
          />
        )}
        {iconBefore || null}
        {children}
        {iconAfter || null}
      </Text>
    )
  }
}

export default withTheme(Button)
