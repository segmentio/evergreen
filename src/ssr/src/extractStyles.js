import React from 'react'
import {
  extractStyles as boxExtractStyles,
  hydrate as boxHydrate
} from 'ui-box'
import { rehydrate } from 'glamor'
import { renderStatic } from 'glamor/server'

const EVERGREEN_SCRIPT_ID = 'evergreen-hydrate'
const isDev = process.env.NODE_ENV === 'development' || !process.env.NODE_ENV

/**
 * You shouldn't have to manually run this.
 * This is mainly an export for testing purposes.
 */
export function hydrate(hydration) {
  if (hydration.uiBoxCache) {
    boxHydrate(hydration.uiBoxCache)
  }

  if (hydration.glamorIds) {
    rehydrate(hydration.glamorIds)
  }
}

function autoHydrate() {
  if (typeof window !== 'undefined' && typeof document !== 'undefined') {
    const hydration = document.getElementById(EVERGREEN_SCRIPT_ID)

    if (hydration) {
      try {
        const hydrationObject = JSON.parse(hydration.innerHTML)
        hydrate(hydrationObject)
      } catch (err) {
        if (isDev) {
          console.log(
            'Evergreen automatic hydration object is invalid JSON',
            err
          )
        }
      }
    }
  }
}

autoHydrate()

export default function extractStyles() {
  const { styles, cache } = boxExtractStyles()
  const { ids, css } = renderStatic(
    () => 'let glamor believe there is some html here'
  )

  const manualHydrateCache = {
    uiBoxCache: cache,
    glamorIds: ids
  }

  const scriptProps = {
    type: 'application/json',
    id: EVERGREEN_SCRIPT_ID,
    dangerouslySetInnerHTML: { __html: JSON.stringify(manualHydrateCache) }
  }

  return {
    css: `${styles} ${css}`,
    manualHydrateCache,
    evergreenHydrateScript: <script {...scriptProps} />
  }
}
