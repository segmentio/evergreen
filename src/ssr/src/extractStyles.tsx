import React from 'react'
import { renderStatic } from 'glamor/server'
import { extractStyles as boxExtractStyles } from 'ui-box'

export default function extractStyles(options = {}) {
  const { cache, styles } = boxExtractStyles()
  const { css, ids } = renderStatic(() => 'let glamor believe there is some html here')

  const evergreenCache = {
    uiBoxCache: cache,
    glamorIds: ids
  }

  const scriptProps = {
    type: 'application/json',
    id: 'evergreen-hydrate',
    dangerouslySetInnerHTML: { __html: JSON.stringify(evergreenCache) }
  }

  if (options.nonce) {
    scriptProps.nonce = options.nonce
  }

  return {
    css: styles + '\n' + css,
    cache: evergreenCache,
    hydrationScript: <script {...scriptProps} />
  }
}
