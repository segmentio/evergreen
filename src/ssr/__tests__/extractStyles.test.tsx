import React from 'react'
import { render } from '@testing-library/react'
import Box from 'ui-box'
import Button from '../../buttons/src/Button'
import extractStyles from '../src/extractStyles'

describe('extractStyles', () => {
  it('returns styles for a Box', () => {
    render(<Box height={11} />)
    const result = extractStyles({ nonce: 'abcd1234' })
    expect(result).toMatchSnapshot()
    expect(result.css).toBeTruthy()
    expect(result.cache.uiBoxCache).toBeTruthy()
    expect(result.cache.glamorIds).toBeTruthy()
    expect(result.hydrationScript).toBeTruthy()
  })

  it('returns styles for a Button', () => {
    render(<Button />)
    const result = extractStyles()
    expect(result).toMatchSnapshot()
  })
})
