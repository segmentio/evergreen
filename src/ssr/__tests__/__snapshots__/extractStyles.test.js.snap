// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractStyles returns styles for a Box 1`] = `
Object {
  "cache": Object {
    "glamorIds": Array [
      "ng405l",
      "fv6wzy",
      "11r1ktn",
    ],
    "uiBoxCache": Array [
      Array [
        "height11",
        "ub-h_11px",
      ],
      Array [
        "boxSizingborder-box",
        "ub-box-szg_border-box",
      ],
    ],
  },
  "css": ".ub-h_11px {height: 11px;}.ub-box-szg_border-box {box-sizing: border-box;}
@-webkit-keyframes loading_ng405l { 
  0% {transform: rotate(0); -webkit-transform: rotate(0);} 
  100% {transform: rotate(360deg); -webkit-transform: rotate(360deg);} 
}@-moz-keyframes loading_ng405l { 
  0% {transform: rotate(0); -webkit-transform: rotate(0);} 
  100% {transform: rotate(360deg); -webkit-transform: rotate(360deg);} 
}@-o-keyframes loading_ng405l { 
  0% {transform: rotate(0); -webkit-transform: rotate(0);} 
  100% {transform: rotate(360deg); -webkit-transform: rotate(360deg);} 
}@keyframes loading_ng405l { 
  0% {transform: rotate(0); -webkit-transform: rotate(0);} 
  100% {transform: rotate(360deg); -webkit-transform: rotate(360deg);} 
}@-webkit-keyframes loading-circle_fv6wzy { 
  0% {stroke-dashoffset: 600;} 
  100% {stroke-dashoffset: 0;} 
}@-moz-keyframes loading-circle_fv6wzy { 
  0% {stroke-dashoffset: 600;} 
  100% {stroke-dashoffset: 0;} 
}@-o-keyframes loading-circle_fv6wzy { 
  0% {stroke-dashoffset: 600;} 
  100% {stroke-dashoffset: 0;} 
}@keyframes loading-circle_fv6wzy { 
  0% {stroke-dashoffset: 600;} 
  100% {stroke-dashoffset: 0;} 
}.css-11r1ktn,[data-css-11r1ktn] {animation: loading_ng405l 2s linear infinite; -webkit-animation: loading_ng405l 2s linear infinite;}",
  "hydrationScript": <script
    dangerouslySetInnerHTML={
      Object {
        "__html": "{\\"uiBoxCache\\":[[\\"height11\\",\\"ub-h_11px\\"],[\\"boxSizingborder-box\\",\\"ub-box-szg_border-box\\"]],\\"glamorIds\\":[\\"ng405l\\",\\"fv6wzy\\",\\"11r1ktn\\"]}",
      }
    }
    id="evergreen-hydrate"
    nonce="abcd1234"
    type="application/json"
  />,
}
`;

exports[`extractStyles returns styles for a Button 1`] = `
Object {
  "cache": Object {
    "glamorIds": Array [
      "ng405l",
      "fv6wzy",
      "11r1ktn",
      "12b25td",
    ],
    "uiBoxCache": Array [
      Array [
        "positionrelative",
        "ub-pst_relative",
      ],
      Array [
        "fontWeight500",
        "ub-f-wght_500",
      ],
      Array [
        "displayinline-flex",
        "ub-dspl_inline-flex",
      ],
      Array [
        "alignItemscenter",
        "ub-algn-itms_center",
      ],
      Array [
        "flexWrapnowrap",
        "ub-flx-wrap_nowrap",
      ],
      Array [
        "justifyContentcenter",
        "ub-just-cnt_center",
      ],
      Array [
        "textDecorationnone",
        "ub-txt-deco_none",
      ],
      Array [
        "verticalAlignmiddle",
        "ub-ver-algn_middle",
      ],
      Array [
        "borderBottom1px solid #c1c4d6",
        "ub-b-btm_1px-solid-c1c4d6",
      ],
      Array [
        "borderLeft1px solid #c1c4d6",
        "ub-b-lft_1px-solid-c1c4d6",
      ],
      Array [
        "borderRight1px solid #c1c4d6",
        "ub-b-rgt_1px-solid-c1c4d6",
      ],
      Array [
        "borderTop1px solid #c1c4d6",
        "ub-b-top_1px-solid-c1c4d6",
      ],
      Array [
        "outlinenone",
        "ub-otln_iu2jf4",
      ],
      Array [
        "userSelectnone",
        "ub-usr-slct_none",
      ],
      Array [
        "cursorpointer",
        "ub-crsr_pointer",
      ],
      Array [
        "whiteSpacenowrap",
        "ub-wht-spc_nowrap",
      ],
      Array [
        "fontFamily\\"SF UI Text\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
        "ub-fnt-fam_b77syt",
      ],
      Array [
        "borderBottomLeftRadius4px",
        "ub-bblr_4px",
      ],
      Array [
        "borderBottomRightRadius4px",
        "ub-bbrr_4px",
      ],
      Array [
        "borderTopLeftRadius4px",
        "ub-btlr_4px",
      ],
      Array [
        "borderTopRightRadius4px",
        "ub-btrr_4px",
      ],
      Array [
        "color#474d66",
        "ub-color_474d66",
      ],
      Array [
        "transitionbox-shadow 80ms ease-in-out",
        "ub-tstn_n1akt6",
      ],
      Array [
        "height32",
        "ub-h_32px",
      ],
      Array [
        "minWidth32",
        "ub-min-w_32px",
      ],
      Array [
        "fontSize12px",
        "ub-fnt-sze_12px",
      ],
      Array [
        "lineHeight32px",
        "ub-ln-ht_32px",
      ],
      Array [
        "paddingLeft16",
        "ub-pl_16px",
      ],
      Array [
        "paddingRight16",
        "ub-pr_16px",
      ],
      Array [
        "backgroundColorwhite",
        "ub-bg-clr_white",
      ],
      Array [
        "boxSizingborder-box",
        "ub-box-szg_border-box",
      ],
    ],
  },
  "css": ".ub-pst_relative {position: relative;}.ub-f-wght_500 {font-weight: 500;}.ub-dspl_inline-flex {display: inline-flex;}.ub-algn-itms_center {align-items: center; -webkit-box-align: center;}.ub-flx-wrap_nowrap {flex-wrap: nowrap; -webkit-box-lines: nowrap;}.ub-just-cnt_center {justify-content: center; -webkit-box-pack: center;}.ub-txt-deco_none {text-decoration: none;}.ub-ver-algn_middle {vertical-align: middle;}.ub-b-btm_1px-solid-c1c4d6 {border-bottom: 1px solid #c1c4d6;}.ub-b-lft_1px-solid-c1c4d6 {border-left: 1px solid #c1c4d6;}.ub-b-rgt_1px-solid-c1c4d6 {border-right: 1px solid #c1c4d6;}.ub-b-top_1px-solid-c1c4d6 {border-top: 1px solid #c1c4d6;}.ub-otln_iu2jf4 {outline: none;}.ub-usr-slct_none {-webkit-user-select: none; -moz-user-select: none; -ms-user-select: none; user-select: none;}.ub-crsr_pointer {cursor: pointer;}.ub-wht-spc_nowrap {white-space: nowrap;}.ub-fnt-fam_b77syt {font-family: \\"SF UI Text\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\";}.ub-bblr_4px {border-bottom-left-radius: 4px;}.ub-bbrr_4px {border-bottom-right-radius: 4px;}.ub-btlr_4px {border-top-left-radius: 4px;}.ub-btrr_4px {border-top-right-radius: 4px;}.ub-color_474d66 {color: #474d66;}.ub-tstn_n1akt6 {transition: box-shadow 80ms ease-in-out;}.ub-h_32px {height: 32px;}.ub-min-w_32px {min-width: 32px;}.ub-fnt-sze_12px {font-size: 12px;}.ub-ln-ht_32px {line-height: 32px;}.ub-pl_16px {padding-left: 16px;}.ub-pr_16px {padding-right: 16px;}.ub-bg-clr_white {background-color: white;}.ub-box-szg_border-box {box-sizing: border-box;}
@-webkit-keyframes loading_ng405l { 
  0% {transform: rotate(0); -webkit-transform: rotate(0);} 
  100% {transform: rotate(360deg); -webkit-transform: rotate(360deg);} 
}@-moz-keyframes loading_ng405l { 
  0% {transform: rotate(0); -webkit-transform: rotate(0);} 
  100% {transform: rotate(360deg); -webkit-transform: rotate(360deg);} 
}@-o-keyframes loading_ng405l { 
  0% {transform: rotate(0); -webkit-transform: rotate(0);} 
  100% {transform: rotate(360deg); -webkit-transform: rotate(360deg);} 
}@keyframes loading_ng405l { 
  0% {transform: rotate(0); -webkit-transform: rotate(0);} 
  100% {transform: rotate(360deg); -webkit-transform: rotate(360deg);} 
}@-webkit-keyframes loading-circle_fv6wzy { 
  0% {stroke-dashoffset: 600;} 
  100% {stroke-dashoffset: 0;} 
}@-moz-keyframes loading-circle_fv6wzy { 
  0% {stroke-dashoffset: 600;} 
  100% {stroke-dashoffset: 0;} 
}@-o-keyframes loading-circle_fv6wzy { 
  0% {stroke-dashoffset: 600;} 
  100% {stroke-dashoffset: 0;} 
}@keyframes loading-circle_fv6wzy { 
  0% {stroke-dashoffset: 600;} 
  100% {stroke-dashoffset: 0;} 
}.css-11r1ktn,[data-css-11r1ktn] {animation: loading_ng405l 2s linear infinite; -webkit-animation: loading_ng405l 2s linear infinite;}.css-12b25td,[data-css-12b25td] {-webkit-font-smoothing: antialiased; -webkit-appearance: none; -moz-appearance: none;}.css-12b25td::-moz-focus-inner ,[data-css-12b25td]::-moz-focus-inner {border: 0;}.css-12b25td:not([disabled]):focus,[data-css-12b25td]:not([disabled]):focus {box-shadow: 0 0 0 2px #D6E0FF;}.css-12b25td[disabled],[data-css-12b25td][disabled] {cursor: not-allowed; pointer-events: none; color: #c1c4d6; border-color: #E6E8F0;}.css-12b25td:not([disabled]):hover,[data-css-12b25td]:not([disabled]):hover {border: 1px solid #8f95b2; background-color: #FAFBFF;}.css-12b25td:not([disabled]):active,[data-css-12b25td]:not([disabled]):active, .css-12b25td:not([disabled])[aria-expanded=\\"true\\"], [data-css-12b25td]:not([disabled])[aria-expanded=\\"true\\"], .css-12b25td:not([disabled])[data-active], [data-css-12b25td]:not([disabled])[data-active] {background-color: #F4F5F9;}",
  "hydrationScript": <script
    dangerouslySetInnerHTML={
      Object {
        "__html": "{\\"uiBoxCache\\":[[\\"positionrelative\\",\\"ub-pst_relative\\"],[\\"fontWeight500\\",\\"ub-f-wght_500\\"],[\\"displayinline-flex\\",\\"ub-dspl_inline-flex\\"],[\\"alignItemscenter\\",\\"ub-algn-itms_center\\"],[\\"flexWrapnowrap\\",\\"ub-flx-wrap_nowrap\\"],[\\"justifyContentcenter\\",\\"ub-just-cnt_center\\"],[\\"textDecorationnone\\",\\"ub-txt-deco_none\\"],[\\"verticalAlignmiddle\\",\\"ub-ver-algn_middle\\"],[\\"borderBottom1px solid #c1c4d6\\",\\"ub-b-btm_1px-solid-c1c4d6\\"],[\\"borderLeft1px solid #c1c4d6\\",\\"ub-b-lft_1px-solid-c1c4d6\\"],[\\"borderRight1px solid #c1c4d6\\",\\"ub-b-rgt_1px-solid-c1c4d6\\"],[\\"borderTop1px solid #c1c4d6\\",\\"ub-b-top_1px-solid-c1c4d6\\"],[\\"outlinenone\\",\\"ub-otln_iu2jf4\\"],[\\"userSelectnone\\",\\"ub-usr-slct_none\\"],[\\"cursorpointer\\",\\"ub-crsr_pointer\\"],[\\"whiteSpacenowrap\\",\\"ub-wht-spc_nowrap\\"],[\\"fontFamily\\\\\\"SF UI Text\\\\\\", -apple-system, BlinkMacSystemFont, \\\\\\"Segoe UI\\\\\\", Roboto, Helvetica, Arial, sans-serif, \\\\\\"Apple Color Emoji\\\\\\", \\\\\\"Segoe UI Emoji\\\\\\", \\\\\\"Segoe UI Symbol\\\\\\"\\",\\"ub-fnt-fam_b77syt\\"],[\\"borderBottomLeftRadius4px\\",\\"ub-bblr_4px\\"],[\\"borderBottomRightRadius4px\\",\\"ub-bbrr_4px\\"],[\\"borderTopLeftRadius4px\\",\\"ub-btlr_4px\\"],[\\"borderTopRightRadius4px\\",\\"ub-btrr_4px\\"],[\\"color#474d66\\",\\"ub-color_474d66\\"],[\\"transitionbox-shadow 80ms ease-in-out\\",\\"ub-tstn_n1akt6\\"],[\\"height32\\",\\"ub-h_32px\\"],[\\"minWidth32\\",\\"ub-min-w_32px\\"],[\\"fontSize12px\\",\\"ub-fnt-sze_12px\\"],[\\"lineHeight32px\\",\\"ub-ln-ht_32px\\"],[\\"paddingLeft16\\",\\"ub-pl_16px\\"],[\\"paddingRight16\\",\\"ub-pr_16px\\"],[\\"backgroundColorwhite\\",\\"ub-bg-clr_white\\"],[\\"boxSizingborder-box\\",\\"ub-box-szg_border-box\\"]],\\"glamorIds\\":[\\"ng405l\\",\\"fv6wzy\\",\\"11r1ktn\\",\\"12b25td\\"]}",
      }
    }
    id="evergreen-hydrate"
    type="application/json"
  />,
}
`;
